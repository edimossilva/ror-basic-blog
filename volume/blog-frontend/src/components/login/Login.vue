<template lang="">
  <div>
    <p>Login</p>

    <div>
      <label> Username: </label>
      <input type="text" v-model="username" class="Login__username-input-js"> 
    </div>
    
    <div>
      <label> Password: </label>
      <input type="password" v-model="password" class="Login__password-input-js"> 
    </div>

    <button @click="doLogin" class="Login__login-button-js">Login</button>
    <div>
      username: 'registered_user1', password: '111' <br>
      username: 'registered_user2', password: '222' <br>
      username: 'admin_user1',      password: '111' <br>
      username: 'admin_user2',      password: '222' <br>
    </div>
  </div>
</template>

<script>
import authService from "../../services/auth_service";

export default {
  data() {
    return {
      username: "registered_user1",
      password: "111"
    };
  },
  methods: {
    doLogin() {
      const { username, password } = this;
      const { onLoginSuccess, showErrorMessage } = this;

      authService
        .doLogin(username, password)
        .then(onLoginSuccess)
        .catch(showErrorMessage);
    },

    onLoginSuccess({ data: { token, username } }) {
      this.saveToken(token);
      this.saveUsername(username);
      this.redirectToHomeIfLogged();
      console.log("redirecthome");
    },

    saveToken(token) {
      this.$store.commit("saveAuthToken", token);
    },

    saveUsername(username) {
      this.$store.commit("saveUsername", username);
    },

    showErrorMessage(error) {
      // debugger; // eslint-disable-line
      console.log(`error ${error}`);
    },
    redirectToHomeIfLogged() {
      if (this.$store.getters.isLoged) {
        return this.$router.push({ name: "blogs" });
      }
    }
  },
  mounted: function() {
    this.redirectToHomeIfLogged();
  }
};
</script>

<style scoped></style>