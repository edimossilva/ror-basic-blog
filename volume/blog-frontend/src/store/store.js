import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

export const store = new Vuex.Store({
  state: {
    authToken: "",
    username: ""
  },
  mutations: {
    saveAuthToken(state, token) {
      state.authToken = token;
    },
    saveUsername(state, username) {
      state.username = username;
    },
    deleteToken(state) {
      state.authToken = "";
    }
  },
  getters: {
    authToken: state => state.authToken,
    username: state => state.username,
    isLoged: state => !!state.authToken
  }
});