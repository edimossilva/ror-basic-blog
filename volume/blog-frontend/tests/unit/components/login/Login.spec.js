import { mount, config } from '@vue/test-utils'
import Login from '@/components/login/Login'
import { store } from '@/store/store'

config.mocks.$store = store
describe('Login.vue', () => {
  const mockUsername = "mockUsername"
  const mockPassword = "mockPassword"
  const mockToken = "mockToken"

  it('is a valid component', () => {
    const wrapper = mount(Login)

    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  describe('Behaviour', () => {
    describe('When fill username and password', () => {
      const doLoginSpy = jest.fn()
      Login.methods.doLogin = doLoginSpy

      const wrapper = mount(Login)

      beforeEach(() => {
        // fill username 
        const usernameInput = wrapper.find('.Login__username-input-js')
        usernameInput.element.value = mockUsername
        usernameInput.trigger('input')

        //fill password
        const passwordInput = wrapper.find('.Login__password-input-js')
        passwordInput.element.value = mockPassword
        passwordInput.trigger('input')

        //click login button
        const loginButton = wrapper.find('.Login__login-button-js')
        loginButton.trigger('click')
      })

      it('should populate fields and call login method', () => {
        expect(wrapper.vm.username).toEqual(mockUsername)
        expect(wrapper.vm.password).toEqual(mockPassword)
        expect(doLoginSpy).toHaveBeenCalledTimes(1)
      })
    })
  })

  describe('Methods', () => {
    describe('onLoginSuccess', () => {
      const wrapper = mount(Login, {
        store
      })

      beforeEach(() => {
        const response = {
          data: {
            token: mockToken,
            username: mockUsername
          }
        }
        wrapper.vm.onLoginSuccess(response)
      });

      it('should save token and username in store', () => {
        expect(store.state.username).toEqual(mockUsername)
      })
      it('should save token and username in store', () => {
        expect(store.state.authToken).toEqual(mockToken)
      })
    })
  })
})
