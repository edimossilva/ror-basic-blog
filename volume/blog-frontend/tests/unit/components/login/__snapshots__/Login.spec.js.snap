// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.vue is a valid component 1`] = `
<div>
  <p>Login</p>
  <div><label> Username: </label> <input type="text" class="Login__username-input-js"></div>
  <div><label> Password: </label> <input type="password" class="Login__password-input-js"></div> <button class="Login__login-button-js">Login</button>
  <div>
    username: 'registered_user1', password: '111' <br>
    username: 'registered_user2', password: '222' <br>
    username: 'admin_user1', password: '111' <br>
    username: 'admin_user2', password: '222' <br></div>
</div>
`;
